test/unit/deployment/test_extract_validator.py

```diff
@@ -0,0 +1,174 @@
+import contextlib
+import logging
+import pytest
+import re
+import exasol.bucketfs as bfs   # type: ignore
+from pyexasol import ExaConnection
+
+from typing import Any, Dict, List
+from unittest.mock import Mock, call, patch
+from datetime import timedelta
+
+from exasol.python_extension_common.deployment.extract_validator import (
+    ExtractValidator,
+    ExtractException,
+    _udf_name,
+)
+from tenacity import RetryError
+
+LOG = logging.getLogger(__name__)
+
+
+def bucket_path(path: str):
+    bucket_api = bfs.MountedBucket("svc", "bkt")
+    return bfs.path.BucketPath(path, bucket_api=bucket_api)
+
+
+@pytest.fixture
+def archive_bucket_path():
+    return bucket_path("/folder/a.tgz")
+
+
+class ConnectionMock:
+    def __init__(self, spec: Dict[str, Any]):
+        self.spec = spec
+        self.values = iter(())
+
+    def _get_values(self, first_line: str):
+        for regex, values in self.spec.items():
+            if re.match(regex, first_line, re.IGNORECASE):
+                return values() if callable(values) else values
+        LOG.warning(f"ConnectionMock.execute() called with '{first_line[:40]}...'")
+        return ()
+
+    def execute(self, *args, **kwargs):
+        statement = args[0] if len(args) else kwargs["query"]
+        first_line = statement.strip().splitlines()[0]
+        self.values = iter(self._get_values(first_line))
+        return self
+
+    def fetchone(self):
+        return next(self.values)
+
+    def fetchall(self):
+        return [ v for v in self.values ]
+
+
+class Simulator:
+    def __init__(self, nodes: int, udf_results: List[List[any]],
+                 create_script=()):
+        self.create_script = create_script
+        self.nodes = nodes
+        self.udf = Mock(side_effect=udf_results)
+        self.callback = Mock(side_effect = self._callback)
+
+    def _callback(self, n, pending):
+        LOG.debug(f"{len(pending)} of {n} nodes pending: {pending}")
+
+    @property
+    def testee(self):
+        connection = ConnectionMock({
+            r"CREATE .* SCRIPT": self.create_script,
+            r"(CREATE|DROP) ": (),
+            r"SELECT nproc\(\)": [ self.nodes ],
+            r'SELECT .*_manifest_': self.udf,
+        })
+        return ExtractValidator(
+            pyexasol_connection=Mock(execute=connection.execute),
+            timeout=timedelta(seconds=10),
+            interval=timedelta(seconds=1),
+            callback=self.callback,
+        )
+
+
+@contextlib.contextmanager
+def mock_tenacity_wait(*wait_lists: List[int], max: int = 1000):
+    def expand(wait_lists):
```

Comment:
This needs some explanation

Reply:
Thanks! Excellent proposal. See latest push.